### ISSUE TITLE
22.0.5: Empty string user attributes are not stored anymore

### ISSUE THREAD

**Issue by lrozenblyum** [keycloak/keycloak#24389]
### Before reporting an issue

- [X] I have searched existing issues
- [X] I have reproduced the issue with the [latest nightly release](https://github.com/keycloak/keycloak/releases/tag/nightly)

### Area

admin/api

### Describe the bug

User attributes with not-null keys and empty string values aren't stored anymore.

### Version

22.0.5

### Expected behavior

Attributes with not-null keys and empty String values should be stored.

NOTE: as an alternative, if the new behaviour is expected then please document it in the Migration guide or elsewhere.

### Actual behavior

Attributes with not-null keys and empty String values are not stored (or at least not retrievable via API and UI).

### How to Reproduce?

Create a user with an attribute with not-null key and empty String value via Admin API

```
@Test
public void emptyAttributeValueStored() {
	RealmResource realmResource = keycloakIt.realm();
	String username = "attrtest";
	String key = "key";

	UserRepresentation user = new UserRepresentation();
	user.setUsername(username);
	Map<String, List<String>> attributes = new HashMap<>();
	attributes.put(key, Arrays.asList(""));
	user.setAttributes(attributes);

	try (jakarta.ws.rs.core.Response created = realmResource.users().create(user)) {
		assertThat(created.getStatus(), Matchers.equalTo(HttpStatus.SC_CREATED));
	}

	List<UserRepresentation> searchResult = realmResource.users().searchByUsername(username, true);
	assertThat(searchResult, hasSize(1));
	assertThat(searchResult.get(0).getAttributes(), Matchers.aMapWithSize(1));
	assertThat(searchResult.get(0).getAttributes(), Matchers.hasKey(key));
}
```

### Anything else?

The behavior changed somewhere in Keycloak 22.0.4 or 22.0.5 (it worked as expected in 8.0.1 and 22.0.3).

**Comment by rmartinc** [comment]
@lrozenblyum  I think this was done on purpose in https://github.com/keycloak/keycloak/issues/22960 as part of the User Profile effort. This was added in 22.0.5. But note that the change is the empty string `""` is removed, not null as you are mentioning. I have tested and null is not saved since the first quarkus version 17.0.0 (I have not tested in previous versions, so null value is not saved long ago).

@pedroigor Do you want me to add a little upgrade section in 22.0.5 mentioning that empty string is not saved now for user attributes?

**Comment by lrozenblyum** [comment]
Sorry for confusion, you're right @rmartinc - empty strings are not stored anymore.

**Comment by pedroigor** [comment]
@lrozenblyum @rmartinc Yeah, it does make sense to add some notes about this behavior.

**Comment by lrozenblyum** [comment]
@pedroigor @rmartinc as I see existing empty attributes were not cleaned up if I updated the Keycloak from older version, is it ok? 
Should we create a migration that would clean them up at our side or it should be done via Liquibase in Keycloak?
Just worrying if existing Keycloak logic won't break when it sees the empty attribute (which are not allowed to be created anymore)

**Comment by phoenix-ua** [comment]
@pedroigor @rmartinc 

In our project we're going to update Keycloak 8.0.1 to 22.0.5. I came here since I'm also interested in the same question:
> Should we create a migration that would clean them up at our side or it should be done via Liquibase in Keycloak?

BTW, I discovered some peculiarities while testing the aforementioned Keycloak update. The following actions ended up in the appearance of an empty user attribute in Keycloak 22.0.5:

1. Create a `user-attribute-ldap-mapper` for an active LDAP user federation provider, pointing to a LDAP user attribute that consists only of blank chars (e.g. spaces, tabulation,etc `  \t  `). With the next LDAP synchronization cycle such mapper will create an empty user attribute in the destination Keycloak server, having the whitespace in the source attribute totally trimmed.
I expect the same behavior from the case where the source LDAP user attribute is `null` or empty, but I couldn't create such attribute in our LDAP server (slapd).

2. Create an `Attribute Importer` mapper for an active SAML Identity provider and point it to an empty user attribute in an external Identity provider (IdP) server. In our case we used another Keycloak instance as IdP. The first and subsequent logins via SAML will lead to creation/restore of a user attribute with an empty value.

We think it makes no sence to add a migration to delete the empty user attributes that survived the `8.x.x` → `22.x.x` Keycloak update at least due to the findings above. In any case, I would appreciate your feedback on this matter.
